# Si si prueba Probamos a explicar el codigo en un jupyter notebook

## Constructor de permutaciones
Resolveremos cuantas permutaciones sin repeticion posibles hay con los numeros 1,2 y 3:

In [3]:
from itertools import permutations

In [4]:
permutations(range(1,4))

`permutations` nos genera un iterador. Hay varias alternativas para extraer las soluciones.

Usamos el generador de listas `list`

In [13]:
list(permutations(range(1,4)))

[(1, 2, 3), (1, 3, 2), (2, 1, 3), (2, 3, 1), (3, 1, 2), (3, 2, 1)]

O para mayor claridad contruimos un bucle `for`. 

In [14]:
for per in permutations(range(1,4)):
    print(per)

(1, 2, 3)
(1, 3, 2)
(2, 1, 3)
(2, 3, 1)
(3, 1, 2)
(3, 2, 1)


Por algun motivo si probamos a crear un list_comprenhension nos da error **SI NO** añadimos los corchetes:

In [15]:
print(per) for per in permutations(range(1,4))

SyntaxError: invalid syntax (<ipython-input-15-b555e18eda45>, line 1)

Y si los añadimos nos da un output que no queremos

In [16]:
[print(per) for per in permutations(range(1,4))]

(1, 2, 3)
(1, 3, 2)
(2, 1, 3)
(2, 3, 1)
(3, 1, 2)
(3, 2, 1)


[None, None, None, None, None, None]

Para calcular el número de permutaciones posibles, vamos a contruir una sencilla funcion factorial.

## Funcion factorial

In [17]:
import numpy as np
def factorial(n):
     return np.prod(range(1,n+1))

Para las permutaciones sin repeticion de 1,2 y 3, habrá 3! permutaciones.

In [18]:
factorial(3)

6

Como en la ecuacion tendremos los numeros del 1 al 9 sin repetirse, será 9! = 

In [19]:
factorial(9)

362880

## Evaluador de la ecuacion

Tendremos que comprobar 362880 permutaciones. Para ello vamos a definir una funcion que evalue la ecuacion dada una combinacion.

In [20]:
def riddle_equation(coef):
       
    """ Calculamos el resultado de esos coeficientes en la ecuacion """

    try:
        eq = (coef[0] + 
    			  13 
    			  * coef[1]
    			  / coef[2]
    			  + coef[3]
    			  + 12 
    			  * coef[4] 
    			  - coef[5] 
    			  -11 
    			  + coef[6]
    			  * coef[7]
    			  / coef[8]
    			  - 10) # = 66
        return eq
    except ZeroDivisionError:
        return float('NaN')

Utilizamos la funcion con la primera permutacion. 

In [21]:
coef = range(1,10)
print(tuple(coef))
riddle_equation(coef)

(1, 2, 3, 4, 5, 6, 7, 8, 9)


52.88888888888889

Ahora vamos a crear una sencilla funcion para comprobar si el resultado para unso coef dados es 66

In [22]:
def check_coef(coef):
    
    """ Para coef comprobamos que el resultado de la ecuacion es 66 """
    
    if  riddle_equation(coef) == 66:
        return True
    else: 
        return False

## Programa principal

Ya tenemos todos los preparativos. Ahora solo tenemos que 
- generar todas las permutaciones posibles
- Calcular si el resultado es 66
- Almacenar cada solucion. 

In [28]:
def main(): 
    solutions = []
    for per in permutations(range(1,10)):
        coef = tuple(per)
        if check_coef(coef):
            solutions.append(coef)
    return solutions

In [33]:
solution = main()

Ahora os toca a vosotros @HolyTeam!

In [34]:
len(solution)

128

In [37]:
for sol in solution:
    print(sol)

(1, 2, 6, 4, 7, 8, 3, 5, 9)
(1, 2, 6, 4, 7, 8, 5, 3, 9)
(1, 3, 2, 4, 5, 8, 7, 9, 6)
(1, 3, 2, 4, 5, 8, 9, 7, 6)
(1, 3, 2, 9, 5, 6, 4, 7, 8)
(1, 3, 2, 9, 5, 6, 7, 4, 8)
(1, 3, 4, 7, 6, 5, 2, 9, 8)
(1, 3, 4, 7, 6, 5, 9, 2, 8)
(1, 3, 6, 2, 7, 9, 4, 5, 8)
(1, 3, 6, 2, 7, 9, 5, 4, 8)
(1, 3, 9, 4, 7, 8, 2, 5, 6)
(1, 3, 9, 4, 7, 8, 5, 2, 6)
(1, 4, 8, 2, 7, 9, 3, 5, 6)
(1, 4, 8, 2, 7, 9, 5, 3, 6)
(1, 5, 2, 3, 4, 8, 7, 9, 6)
(1, 5, 2, 3, 4, 8, 9, 7, 6)
(1, 5, 2, 8, 4, 7, 3, 9, 6)
(1, 5, 2, 8, 4, 7, 9, 3, 6)
(1, 5, 3, 9, 4, 2, 7, 8, 6)
(1, 5, 3, 9, 4, 2, 8, 7, 6)
(1, 9, 6, 4, 5, 8, 3, 7, 2)
(1, 9, 6, 4, 5, 8, 7, 3, 2)
(1, 9, 6, 7, 5, 2, 3, 4, 8)
(1, 9, 6, 7, 5, 2, 4, 3, 8)
(2, 1, 4, 3, 7, 9, 5, 6, 8)
(2, 1, 4, 3, 7, 9, 6, 5, 8)
(2, 3, 6, 1, 7, 9, 4, 5, 8)
(2, 3, 6, 1, 7, 9, 5, 4, 8)
(2, 4, 8, 1, 7, 9, 3, 5, 6)
(2, 4, 8, 1, 7, 9, 5, 3, 6)
(2, 8, 6, 9, 4, 1, 5, 7, 3)
(2, 8, 6, 9, 4, 1, 7, 5, 3)
(2, 9, 6, 3, 5, 1, 4, 7, 8)
(2, 9, 6, 3, 5, 1, 7, 4, 8)
(3, 1, 4, 2, 7, 9, 5, 6, 8)
(3, 1, 4, 2, 7, 9, 6